File: mindspore_federated/fl_arch/python/common/_fl_context.py
# Copyright 2020 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
"""Context for parameter server training mode"""

import os
from mindspore_federated._mindspore_federated import FLContext
from mindspore_federated.python import log as logger
from mindspore.train.serialization import load_checkpoint, export

_fl_context = None


def fl_context():
    """
    Get the global _fl_context, if it is not created, create a new one.

    Returns:
        _fl_context, the global parameter server training mode context.
    """
    global _fl_context
    if _fl_context is None:
        _fl_context = FLContext.get_instance()
    return _fl_context

_set_fl_context_func_map = {
    "server_mode": fl_context().set_server_mode,
    "ms_role": fl_context().set_ms_role,
    "worker_num": fl_context().set_worker_num,
    "server_num": fl_context().set_server_num,
    "scheduler_ip": fl_context().set_scheduler_ip,
    "scheduler_port": fl_context().set_scheduler_port,
    "fl_server_port": fl_context().set_fl_server_port,
    "enable_fl_client": fl_context().set_fl_client_enable,
    "start_fl_job_threshold": fl_context().set_start_fl_job_threshold,
    "start_fl_job_time_window": fl_context().set_start_fl_job_time_window,
    "update_model_ratio": fl_context().set_update_model_ratio,
    "update_model_time_window": fl_context().set_update_model_time_window,
    "share_secrets_ratio": fl_context().set_share_secrets_ratio,
    "cipher_time_window": fl_context().set_cipher_time_window,
    "reconstruct_secrets_threshold": fl_context().set_reconstruct_secrets_threshold,
    "fl_name": fl_context().set_fl_name,
    "fl_iteration_num": fl_context().set_fl_iteration_num,
    "client_epoch_num": fl_context().set_client_epoch_num,
    "client_batch_size": fl_context().set_client_batch_size,
    "client_learning_rate": fl_context().set_client_learning_rate,
    "worker_step_num_per_iteration": fl_context().set_worker_step_num_per_iteration,
    "root_first_ca_path": fl_context().set_root_first_ca_path,
    "root_second_ca_path": fl_context().set_root_second_ca_path,
    "pki_verify": fl_context().set_pki_verify,
    "equip_crl_path": fl_context().set_equip_crl_path,
    "replay_attack_time_diff": fl_context().set_replay_attack_time_diff,
    "enable_ssl": fl_context().set_enable_ssl,
    "client_password": fl_context().set_client_password,
    "server_password": fl_context().set_server_password,
    "scheduler_manage_port": fl_context().set_scheduler_manage_port,
    "config_file_path": fl_context().set_config_file_path,
    "dp_eps": fl_context().set_dp_eps,
    "dp_delta": fl_context().set_dp_delta,
    "dp_norm_clip": fl_context().set_dp_norm_clip,
    "encrypt_type": fl_context().set_encrypt_type,
    "http_url_prefix": fl_context().set_http_url_prefix,
    "global_iteration_time_window": fl_context().set_global_iteration_time_window,
    "sign_k": fl_context().set_sign_k,
    "sign_eps": fl_context().set_sign_eps,
    "sign_thr_ratio": fl_context().set_sign_thr_ratio,
    "sign_global_lr": fl_context().set_sign_global_lr,
    "sign_dim_out": fl_context().set_sign_dim_out,
    "checkpoint_dir": fl_context().set_checkpoint_dir,
    "upload_compress_type": fl_context().set_upload_compress_type,
    "upload_sparse_rate": fl_context().set_upload_sparse_rate,
    "download_compress_type": fl_context().set_download_compress_type,
    "instance_name": fl_context().set_instance_name,
    "participation_time_level": fl_context().set_participation_time_level,
    "continuous_failure_times": fl_context().set_continuous_failure_times,
    "feature_maps": fl_context().set_feature_maps,
}

_get_fl_context_func_map = {
    "server_mode": fl_context().server_mode,
    "ms_role": fl_context().ms_role,
    "worker_num": fl_context().worker_num,
    "server_num": fl_context().server_num,
    "scheduler_ip": fl_context().scheduler_ip,
    "scheduler_port": fl_context().scheduler_port,
    "fl_server_port": fl_context().fl_server_port,
    "enable_fl_client": fl_context().fl_client_enable,
    "start_fl_job_threshold": fl_context().start_fl_job_threshold,
    "start_fl_job_time_window": fl_context().start_fl_job_time_window,
    "update_model_ratio": fl_context().update_model_ratio,
    "update_model_time_window": fl_context().update_model_time_window,
    "share_secrets_ratio": fl_context().share_secrets_ratio,
    "cipher_time_window": fl_context().cipher_time_window,
    "reconstruct_secrets_threshold": fl_context().reconstruct_secrets_threshold,
    "fl_name": fl_context().fl_name,
    "fl_iteration_num": fl_context().fl_iteration_num,
    "client_epoch_num": fl_context().client_epoch_num,
    "client_batch_size": fl_context().client_batch_size,
    "client_learning_rate": fl_context().client_learning_rate,
    "worker_step_num_per_iteration": fl_context().worker_step_num_per_iteration,
    "dp_eps": fl_context().dp_eps,
    "dp_delta": fl_context().dp_delta,
    "dp_norm_clip": fl_context().dp_norm_clip,
    "encrypt_type": fl_context().encrypt_type,
    "root_first_ca_path": fl_context().root_first_ca_path,
    "root_second_ca_path": fl_context().root_second_ca_path,
    "pki_verify": fl_context().pki_verify,
    "equip_crl_path": fl_context().equip_crl_path,
    "replay_attack_time_diff": fl_context().replay_attack_time_diff,
    "enable_ssl": fl_context().enable_ssl,
    "client_password": fl_context().client_password,
    "server_password": fl_context().server_password,
    "scheduler_manage_port": fl_context().scheduler_manage_port,
    "config_file_path": fl_context().config_file_path,
    "http_url_prefix": fl_context().http_url_prefix,
    "global_iteration_time_window": fl_context().global_iteration_time_window,
    "sign_k": fl_context().sign_k,
    "sign_eps": fl_context().sign_eps,
    "sign_thr_ratio": fl_context().sign_thr_ratio,
    "sign_global_lr": fl_context().sign_global_lr,
    "sign_dim_out": fl_context().sign_dim_out,
    "checkpoint_dir": fl_context().checkpoint_dir,
    "upload_compress_type": fl_context().upload_compress_type,
    "upload_sparse_rate": fl_context().upload_sparse_rate,
    "download_compress_type": fl_context().download_compress_type,
    "instance_name": fl_context().instance_name,
    "participation_time_level": fl_context().participation_time_level,
    "continuous_failure_times": fl_context().continuous_failure_times,
}

_check_positive_int_keys = ["server_num", "scheduler_port", "fl_server_port",
                            "start_fl_job_threshold", "start_fl_job_time_window", "update_model_time_window",
                            "fl_iteration_num", "client_epoch_num", "client_batch_size", "cipher_time_window",
                            "reconstruct_secrets_threshold"]

_check_non_negative_int_keys = ["worker_num"]

_check_positive_float_keys = ["update_model_ratio", "client_learning_rate"]

_check_port_keys = ["scheduler_port", "fl_server_port"]

_check_string_keys = {
    "upload_compress_type": ["NO_COMPRESS", "DIFF_SPARSE_QUANT"],
    "download_compress_type": ["NO_COMPRESS", "QUANT"],
}

def _set_fl_context(kwargs):
    """
    Set parameter server training mode context.

    Note:
        Some other environment variables should also be set for parameter server training mode.
        These environment variables are listed below:

        .. code-block::

            MS_SERVER_NUM  # Server number
            MS_WORKER_NUM  # Worker number
            MS_SCHED_HOST  # Scheduler IP address
            MS_SCHED_PORT  # Scheduler port
            MS_ROLE        # The role of this process:
                           # MS_SCHED represents the scheduler,
                           # MS_WORKER represents the worker,
                           # MS_PSERVER represents the Server


    Args:
        enable_ps (bool): Whether to enable parameter server training mode.
                          Only after enable_ps is set True, the environment variables will be effective.
                          Default: False.
        config_file_path (string): Configuration file path used by recovery. Default: ''.
        scheduler_manage_port (int): scheduler manage port used to scale out/in. Default: 11202.
        enable_ssl (bool): Set PS SSL mode enabled or disabled. Default: False.
        client_password (str): Password to decrypt the secret key stored in the client certificate. Default: ''.
        server_password (str): Password to decrypt the secret key stored in the server certificate. Default: ''.

    Raises:
        ValueError: If input key is not the attribute in parameter server training mode context.

    Examples:
        >>> context.set_fl_context(enable_ps=True, enable_ssl=True, client_password='123456', server_password='123456')
    """
    kwargs = _check_conflict_value(kwargs)
    for key, value in kwargs.items():
        if key not in _set_fl_context_func_map:
            raise ValueError("Set PS context keyword %s is not recognized!" % key)
        logger.info("FL context key is {}, value is {}".format(key, value))
        if key == "checkpoint_dir" and (kwargs["ms_role"] == "MS_SERVER" or kwargs["ms_role"] == "MS_WORKER"):
            checkpoint_dir = kwargs["checkpoint_dir"]
            if (os.path.exists(checkpoint_dir)):
                for checkpoint_file in os.listdir(checkpoint_dir):
                    logger.info("Checkpoint file is {}".format(checkpoint_file))
                    file_prefix = kwargs["fl_name"] + "_iteration"
                    if checkpoint_file.startswith(file_prefix):
                        checkpoint_file_path = os.path.join(checkpoint_dir, checkpoint_file)
                        param_dict = load_checkpoint(checkpoint_file_path)
                        logger.info("Param_dict is {}".format(param_dict))

            weight_fullnames = list()
            weight_datas = list()
            weight_shapes = list()
            weight_types = list()
            for param_name, param_value in param_dict.items():
                weight_fullnames.append(param_name)
                weight_np = param_value.asnumpy()
                weight_shapes.append(list(weight_np.shape))
                weight_datas.append(weight_np.reshape(-1).tolist())
                weight_types.append(str(weight_np.dtype).title())
                logger.info("Weight fullname is {}".format(param_name))
            set_func = _set_fl_context_func_map["feature_maps"]
            set_func(weight_fullnames, weight_datas, weight_shapes, weight_types)
        set_func = _set_fl_context_func_map[key]
        set_func(value)


def _get_fl_context(attr_key):
    """
    Get parameter server training mode context attribute value according to the key.

    Args:
        attr_key (str): The key of the attribute.

    Returns:
        Returns attribute value according to the key.

    Raises:
        ValueError: If input key is not attribute in auto parallel context.
    """
    if attr_key not in _get_fl_context_func_map:
        raise ValueError("Get PS context keyword %s is not recognized!" % attr_key)
    get_func = _get_fl_context_func_map[attr_key]
    value = get_func()
    return value


def _reset_fl_context():
    """
    Reset parameter server training mode context attributes to the default values:

    - enable_ps: False.
    """
    fl_context().reset()


def _is_role_worker():
    return fl_context().is_worker()


def _is_role_pserver():
    return fl_context().is_server()


def _is_role_sched():
    return fl_context().is_scheduler()


def _insert_hash_table_size(name, cache_vocab_size, embedding_size, vocab_size):
    fl_context().insert_hash_table_size(name, cache_vocab_size, embedding_size, vocab_size)


def _reinsert_hash_table_size(new_name, cur_name, cache_vocab_size, embedding_size):
    fl_context().reinsert_hash_table_size(new_name, cur_name, cache_vocab_size, embedding_size)


def _insert_weight_init_info(name, global_seed, op_seed):
    fl_context().insert_weight_init_info(name, global_seed, op_seed)


def _insert_accumu_init_info(name, init_val):
    fl_context().insert_accumu_init_info(name, init_val)


def _check_conflict_value(kwargs):
    if "upload_compress_type" in kwargs and "encrypt_type" in kwargs:
        if kwargs["upload_compress_type"] != "NO_COMPRESS" and kwargs["encrypt_type"] in ("SIGNDS", "PW_ENCRYPT"):
            logger.warning("The '{}' and '{}' are conflicted, and in '{}' mode the"
                           " 'upload_compress_type' will be 'NO_COMPRESS'".format(kwargs["encrypt_type"],
                                                                                  kwargs["upload_compress_type"],
                                                                                  kwargs["encrypt_type"]))
            kwargs["upload_compress_type"] = "NO_COMPRESS"
    return kwargs

File: mindspore_federated/fl_arch/ccsrc/python/fl_context.cc
/**
 * Copyright 2020 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "python/fl_context.h"
#include "common/utils/log_adapter.h"

namespace mindspore {
namespace fl {
std::shared_ptr<FLContext> FLContext::instance() {
  static std::shared_ptr<FLContext> instance = nullptr;
  if (instance == nullptr) {
    instance.reset(new FLContext());
  }
  return instance;
}

void FLContext::Reset() {
  is_worker_ = false;
  is_server_ = false;
  is_sched_ = false;
}

std::string FLContext::ms_role() const {
  if (is_worker_) {
    return kEnvRoleOfWorker;
  } else if (is_server_) {
    return kEnvRoleOfServer;
  } else if (is_sched_) {
    return kEnvRoleOfScheduler;
  } else {
    MS_LOG(EXCEPTION) << "MS role is disabled.";
  }
}

bool FLContext::is_worker() const { return is_worker_; }

bool FLContext::is_server() const { return is_server_; }

bool FLContext::is_scheduler() const { return is_sched_; }

void FLContext::set_server_mode(const std::string &server_mode) {
  if (server_mode != kServerModeFL && server_mode != kServerModeHybrid) {
    MS_LOG(EXCEPTION) << server_mode << " is invalid. Server mode must be " << kServerModeFL << " or "
                      << kServerModeHybrid;
    return;
  }
  MS_LOG(INFO) << "Server mode: " << server_mode << " is used for Server and Worker. Scheduler will ignore it.";
  server_mode_ = server_mode;
}

const std::string &FLContext::server_mode() const { return server_mode_; }

void FLContext::set_encrypt_type(const std::string &encrypt_type) {
  if (encrypt_type != kNotEncryptType && encrypt_type != kDPEncryptType && encrypt_type != kPWEncryptType &&
      encrypt_type != kStablePWEncryptType && encrypt_type != kDSEncryptType) {
    MS_LOG(WARNING) << encrypt_type << " is invalid. Encrypt type must be " << kNotEncryptType << " or "
                    << kDPEncryptType << " or " << kPWEncryptType << " or " << kStablePWEncryptType << " or "
                    << kDSEncryptType << ", DP scheme is used by default.";
    encrypt_type_ = kDPEncryptType;
  } else {
    encrypt_type_ = encrypt_type;
  }
}

const std::string &FLContext::encrypt_type() const { return encrypt_type_; }

void FLContext::set_dp_eps(float dp_eps) {
  if (dp_eps > 0) {
    dp_eps_ = dp_eps;
  } else {
    MS_LOG(WARNING) << dp_eps << " is invalid, dp_eps must be larger than 0, 50 is used by default.";
    float dp_eps_default = 50;
    dp_eps_ = dp_eps_default;
  }
}

float FLContext::dp_eps() const { return dp_eps_; }

void FLContext::set_dp_delta(float dp_delta) {
  if (dp_delta > 0 && dp_delta < 1) {
    dp_delta_ = dp_delta;
  } else {
    MS_LOG(WARNING) << dp_delta << " is invalid, dp_delta must be in range of (0, 1), 0.01 is used by default.";
    float dp_delta_default = 0.01;
    dp_delta_ = dp_delta_default;
  }
}
float FLContext::dp_delta() const { return dp_delta_; }

void FLContext::set_dp_norm_clip(float dp_norm_clip) {
  if (dp_norm_clip > 0) {
    dp_norm_clip_ = dp_norm_clip;
  } else {
    MS_LOG(WARNING) << dp_norm_clip << " is invalid, dp_norm_clip must be larger than 0, 1 is used by default.";
    float dp_norm_clip_default = 1;
    dp_norm_clip_ = dp_norm_clip_default;
  }
}
float FLContext::dp_norm_clip() const { return dp_norm_clip_; }

void FLContext::set_sign_k(float sign_k) {
  float sign_k_upper = 0.25;
  if (sign_k > 0 && sign_k <= sign_k_upper) {
    sign_k_ = sign_k;
  } else {
    MS_LOG(WARNING) << sign_k << " is invalid, sign_k must be in range of (0, 0.25], 0.01 is used by default.";
    float sign_k_default = 0.01;
    sign_k_ = sign_k_default;
  }
}
float FLContext::sign_k() const { return sign_k_; }

void FLContext::set_sign_eps(float sign_eps) {
  float sign_eps_upper = 100;
  if (sign_eps > 0 && sign_eps <= sign_eps_upper) {
    sign_eps_ = sign_eps;
  } else {
    MS_LOG(WARNING) << sign_eps << " is invalid, sign_eps must be in range of (0, 100], 100 is used by default.";
    float sign_eps_default = 100;
    sign_eps_ = sign_eps_default;
  }
}
float FLContext::sign_eps() const { return sign_eps_; }

void FLContext::set_sign_thr_ratio(float sign_thr_ratio) {
  float sign_thr_ratio_bound = 0.5;
  if (sign_thr_ratio >= sign_thr_ratio_bound && sign_thr_ratio <= 1) {
    sign_thr_ratio_ = sign_thr_ratio;
  } else {
    MS_LOG(WARNING) << sign_thr_ratio
                    << " is invalid, sign_thr_ratio must be in range of [0.5, 1], 0.6 is used by default.";
    float sign_thr_ratio_default = 0.6;
    sign_thr_ratio_ = sign_thr_ratio_default;
  }
}
float FLContext::sign_thr_ratio() const { return sign_thr_ratio_; }

void FLContext::set_sign_global_lr(float sign_global_lr) {
  if (sign_global_lr > 0) {
    sign_global_lr_ = sign_global_lr;
  } else {
    MS_LOG(WARNING) << sign_global_lr << " is invalid, sign_global_lr must be larger than 0, 1 is used by default.";
    float sign_global_lr_default = 1;
    sign_global_lr_ = sign_global_lr_default;
  }
}
float FLContext::sign_global_lr() const { return sign_global_lr_; }

void FLContext::set_sign_dim_out(int sign_dim_out) {
  int sign_dim_out_upper = 50;
  if (sign_dim_out >= 0 && sign_dim_out <= sign_dim_out_upper) {
    sign_dim_out_ = sign_dim_out;
  } else {
    MS_LOG(WARNING) << sign_dim_out << " is invalid, sign_dim_out must be in range of [0, 50], 0 is used by default.";
    sign_dim_out_ = 0;
  }
}
int FLContext::sign_dim_out() const { return sign_dim_out_; }

void FLContext::set_ms_role(const std::string &role) {
  if (role == kEnvRoleOfWorker) {
    is_worker_ = true;
  } else if (role == kEnvRoleOfServer) {
    is_server_ = true;
  } else if (role == kEnvRoleOfScheduler) {
    is_sched_ = true;
  } else {
    MS_LOG(EXCEPTION) << "ms_role " << role << " is invalid.";
    return;
  }
  MS_LOG(INFO) << "MS_ROLE of this node is " << role;
  role_ = role;
}

void FLContext::set_worker_num(uint32_t worker_num) {
  // Hybrid training mode only supports one worker for now.
  if (server_mode_ == kServerModeHybrid && worker_num != 1) {
    MS_LOG(EXCEPTION) << "The worker number should be set to 1 in hybrid training mode.";
    return;
  }
  worker_num_ = worker_num;
}
uint32_t FLContext::initial_worker_num() const { return worker_num_; }

void FLContext::set_server_num(uint32_t server_num) { server_num_ = server_num; }
uint32_t FLContext::initial_server_num() const { return server_num_; }

void FLContext::set_scheduler_ip(const std::string &sched_ip) {
  if (sched_ip.length() > kLength) {
    MS_LOG(EXCEPTION) << "The scheduler ip's length can not exceed " << kLength;
  }
  scheduler_ip_ = sched_ip;
}

std::string FLContext::scheduler_ip() const { return scheduler_ip_; }

void FLContext::set_scheduler_port(uint16_t sched_port) {
  if (sched_port > kMaxPort) {
    MS_LOG(EXCEPTION) << "The port: " << sched_port << " is illegal.";
  }
  scheduler_port_ = sched_port;
}

uint16_t FLContext::scheduler_port() const { return scheduler_port_; }

void FLContext::GenerateResetterRound() {
  uint32_t binary_server_context = 0;
  bool is_federated_learning_mode = false;
  bool is_mixed_training_mode = false;
  bool is_pairwise_encrypt = (encrypt_type_ == kPWEncryptType);

  if (server_mode_ == kServerModeFL) {
    is_federated_learning_mode = true;
  } else if (server_mode_ == kServerModeHybrid) {
    is_mixed_training_mode = true;
  } else {
    MS_LOG(EXCEPTION) << server_mode_ << " is invalid. Server mode must be " << " or " << kServerModeFL
                      << " or " << kServerModeHybrid;
    return;
  }
  const int training_mode_offset = 2;
  const int pairwise_encrypt_offset = 3;
  binary_server_context = ((unsigned int)is_federated_learning_mode << 1) |
                          ((unsigned int)is_mixed_training_mode << training_mode_offset) |
                          ((unsigned int)is_pairwise_encrypt << pairwise_encrypt_offset);
  if (kServerContextToResetRoundMap.count(binary_server_context) == 0) {
    resetter_round_ = ResetterRound::kNoNeedToReset;
  } else {
    resetter_round_ = kServerContextToResetRoundMap.at(binary_server_context);
  }
  MS_LOG(INFO) << "Server context is " << binary_server_context << ". Resetter round is " << resetter_round_;
  return;
}

ResetterRound FLContext::resetter_round() const { return resetter_round_; }

void FLContext::set_fl_server_port(uint16_t fl_server_port) { fl_server_port_ = fl_server_port; }

uint16_t FLContext::fl_server_port() const { return fl_server_port_; }

void FLContext::set_fl_client_enable(bool enabled) { fl_client_enable_ = enabled; }

bool FLContext::fl_client_enable() const { return fl_client_enable_; }

void FLContext::set_start_fl_job_threshold(uint64_t start_fl_job_threshold) {
  start_fl_job_threshold_ = start_fl_job_threshold;
}

uint64_t FLContext::start_fl_job_threshold() const { return start_fl_job_threshold_; }

void FLContext::set_start_fl_job_time_window(uint64_t start_fl_job_time_window) {
  start_fl_job_time_window_ = start_fl_job_time_window;
}

uint64_t FLContext::start_fl_job_time_window() const { return start_fl_job_time_window_; }

void FLContext::set_update_model_ratio(float update_model_ratio) {
  if (update_model_ratio > 1.0) {
    MS_LOG(EXCEPTION) << "update_model_ratio must be between 0 and 1.";
    return;
  }
  update_model_ratio_ = update_model_ratio;
}

float FLContext::update_model_ratio() const { return update_model_ratio_; }

void FLContext::set_update_model_time_window(uint64_t update_model_time_window) {
  update_model_time_window_ = update_model_time_window;
}

uint64_t FLContext::update_model_time_window() const { return update_model_time_window_; }

void FLContext::set_share_secrets_ratio(float share_secrets_ratio) {
  if (share_secrets_ratio > 0 && share_secrets_ratio <= 1) {
    share_secrets_ratio_ = share_secrets_ratio;
  } else {
    MS_LOG(EXCEPTION) << share_secrets_ratio << " is invalid, share_secrets_ratio must be in range of (0, 1].";
    return;
  }
}

float FLContext::share_secrets_ratio() const { return share_secrets_ratio_; }

void FLContext::set_cipher_time_window(uint64_t cipher_time_window) {
  if (cipher_time_window_ < 0) {
    MS_LOG(EXCEPTION) << "cipher_time_window should not be less than 0.";
    return;
  }
  cipher_time_window_ = cipher_time_window;
}

uint64_t FLContext::cipher_time_window() const { return cipher_time_window_; }

void FLContext::set_reconstruct_secrets_threshold(uint64_t reconstruct_secrets_threshold) {
  if (reconstruct_secrets_threshold == 0) {
    MS_LOG(EXCEPTION) << "reconstruct_secrets_threshold should be positive.";
    return;
  }
  reconstruct_secrets_threshold_ = reconstruct_secrets_threshold;
}

uint64_t FLContext::reconstruct_secrets_threshold() const { return reconstruct_secrets_threshold_; }

void FLContext::set_fl_name(const std::string &fl_name) { fl_name_ = fl_name; }

const std::string &FLContext::fl_name() const { return fl_name_; }

void FLContext::set_fl_iteration_num(uint64_t fl_iteration_num) { fl_iteration_num_ = fl_iteration_num; }

uint64_t FLContext::fl_iteration_num() const { return fl_iteration_num_; }

void FLContext::set_client_epoch_num(uint64_t client_epoch_num) { client_epoch_num_ = client_epoch_num; }

uint64_t FLContext::client_epoch_num() const { return client_epoch_num_; }

void FLContext::set_client_batch_size(uint64_t client_batch_size) { client_batch_size_ = client_batch_size; }

uint64_t FLContext::client_batch_size() const { return client_batch_size_; }

void FLContext::set_client_learning_rate(float client_learning_rate) { client_learning_rate_ = client_learning_rate; }

float FLContext::client_learning_rate() const { return client_learning_rate_; }

void FLContext::set_secure_aggregation(bool secure_aggregation) { secure_aggregation_ = secure_aggregation; }

bool FLContext::secure_aggregation() const { return secure_aggregation_; }

fl::core::ClusterConfig &FLContext::cluster_config() {
  if (cluster_config_ == nullptr) {
    cluster_config_ = std::make_unique<fl::core::ClusterConfig>(worker_num_, server_num_, scheduler_ip_, scheduler_port_);
  }
  return *cluster_config_;
}

void FLContext::set_root_first_ca_path(const std::string &root_first_ca_path) {
  root_first_ca_path_ = root_first_ca_path;
}
void FLContext::set_root_second_ca_path(const std::string &root_second_ca_path) {
  root_second_ca_path_ = root_second_ca_path;
}

std::string FLContext::root_first_ca_path() const { return root_first_ca_path_; }
std::string FLContext::root_second_ca_path() const { return root_second_ca_path_; }

void FLContext::set_pki_verify(bool pki_verify) { pki_verify_ = pki_verify; }
bool FLContext::pki_verify() const { return pki_verify_; }

void FLContext::set_replay_attack_time_diff(uint64_t replay_attack_time_diff) {
  replay_attack_time_diff_ = replay_attack_time_diff;
}

uint64_t FLContext::replay_attack_time_diff() const { return replay_attack_time_diff_; }

std::string FLContext::equip_crl_path() const { return equip_crl_path_; }

void FLContext::set_equip_crl_path(const std::string &equip_crl_path) { equip_crl_path_ = equip_crl_path; }

void FLContext::set_scheduler_manage_port(uint16_t sched_port) {
  if (sched_port > kMaxPort) {
    MS_LOG(EXCEPTION) << "The port << " << sched_port << " is illegal.";
  }
  scheduler_manage_port_ = sched_port;
}

uint16_t FLContext::scheduler_manage_port() const { return scheduler_manage_port_; }

void FLContext::set_config_file_path(const std::string &path) { config_file_path_ = path; }

std::string FLContext::config_file_path() const { return config_file_path_; }

void FLContext::set_node_id(const std::string &node_id) {
  if (node_id.length() > kLength) {
    MS_LOG(EXCEPTION) << "The node id length can not exceed " << kLength;
  }
  node_id_ = node_id;
}

const std::string &FLContext::node_id() const { return node_id_; }

bool FLContext::enable_ssl() const { return enable_ssl_; }

void FLContext::set_enable_ssl(bool enabled) { enable_ssl_ = enabled; }

std::string FLContext::client_password() const { return client_password_; }
void FLContext::set_client_password(const std::string &password) { client_password_ = password; }

std::string FLContext::server_password() const { return server_password_; }
void FLContext::set_server_password(const std::string &password) { server_password_ = password; }

std::string FLContext::http_url_prefix() const { return http_url_prefix_; }

void FLContext::set_http_url_prefix(const std::string &http_url_prefix) { http_url_prefix_ = http_url_prefix; }

void FLContext::set_global_iteration_time_window(const uint64_t &global_iteration_time_window) {
  global_iteration_time_window_ = global_iteration_time_window;
}

uint64_t FLContext::global_iteration_time_window() const { return global_iteration_time_window_; }

void FLContext::set_upload_compress_type(const std::string &upload_compress_type) {
  upload_compress_type_ = upload_compress_type;
}
std::string FLContext::upload_compress_type() const { return upload_compress_type_; }

void FLContext::set_upload_sparse_rate(float upload_sparse_rate) { upload_sparse_rate_ = upload_sparse_rate; }
float FLContext::upload_sparse_rate() const { return upload_sparse_rate_; }

void FLContext::set_download_compress_type(const std::string &download_compress_type) {
  download_compress_type_ = download_compress_type;
}
std::string FLContext::download_compress_type() const { return download_compress_type_; }

std::string FLContext::checkpoint_dir() const { return checkpoint_dir_; }

void FLContext::set_checkpoint_dir(const std::string &checkpoint_dir) { checkpoint_dir_ = checkpoint_dir; }

void FLContext::set_instance_name(const std::string &instance_name) { instance_name_ = instance_name; }

const std::string &FLContext::instance_name() const { return instance_name_; }

void FLContext::set_participation_time_level(const std::string &participation_time_level) {
  participation_time_level_ = participation_time_level;
}

const std::string &FLContext::participation_time_level() { return participation_time_level_; }

void FLContext::set_continuous_failure_times(uint32_t continuous_failure_times) {
  continuous_failure_times_ = continuous_failure_times;
}

uint32_t FLContext::continuous_failure_times() { return continuous_failure_times_; }

void FLContext::set_feature_maps(std::vector<std::string> weight_fullnames, std::vector<std::vector<float>> weight_datas,
                                 std::vector<std::vector<size_t>> weight_shapes, std::vector<std::string> weight_types) {
  for (size_t i = 0; i < weight_fullnames.size(); i++) {
    std::string weight_fullname = weight_fullnames[i];
    std::vector<size_t> weight_shape = weight_shapes[i];
    std::string weight_type = weight_types[i];
    size_t weight_size = std::accumulate(weight_shape.begin(), weight_shape.end(), sizeof(float), std::multiplies<size_t>());
    std::vector<float> weight_data = weight_datas[i];

    Feature feature;
    feature.weight_shape = weight_shape;
    feature.weight_type = weight_type;
    feature.weight_size = weight_size;
    feature.weight_data = weight_data;

    feature_maps_[weight_fullname] = feature;
    MS_LOG(INFO) << "Weight fullname is : " << weight_fullname << ", weight size is " << weight_size
                 << ", weight shape is " << weight_shape << ", weight type is " << weight_type;

    MS_LOG(DEBUG) << "Weight data is:" << weight_data[0] << " " << weight_data[1] << " " << weight_data[2] << " ";
  }
}
std::map<std::string, Feature> &FLContext::feature_maps() {
  return feature_maps_;
}
}  // namespace fl
}  // namespace mindspore
